src/cli/subcommand/split_receipt.rs

```diff
@@ -0,0 +1,478 @@
+                    phase2_plan
+                        .actions
+                        .push(action_obj["create_volume"].clone());
+                    if !action_obj["encrypt_volume"].is_null() {
+                        phase2_plan
+                            .actions
+                            .push(action_obj["encrypt_volume"].clone());
+                    }
+                    phase2_plan
+                        .actions
+                        .push(action_obj["unmount_volume"].clone());
+                }
+
+                {
+                    action_obj["create_volume"]["state"] = skipped_json_value.clone();
+                    if !action_obj["encrypt_volume"].is_null() {
+                        action_obj["encrypt_volume"]["state"] = skipped_json_value.clone();
+                    }
+                    action_obj["unmount_volume"]["state"] = skipped_json_value.clone();
+                }
+            },
+            // CreateDeterminateNixVolume; macOS-only
+            "create_determinate_nix_volume" => {
+                tracing::debug!("Marking create_volume, encrypt_volume, unmount_volume as skipped so we don't undo it until phase 2");
+
+                {
+                    phase2_plan
+                        .actions
+                        .push(action_obj["create_volume"].clone());
+                    phase2_plan
+                        .actions
+                        .push(action_obj["encrypt_volume"].clone());
+                    // NOTE(cole-h): this action will ~always be at the "Progress" phase because we
+                    // expect it to fail, and once it fails, it exits early and doesn't set the
+                    // completed status
+                    phase2_plan
+                        .actions
+                        .push(action_obj["unmount_volume"].clone());
+                }
+
+                {
+                    action_obj["create_volume"]["state"] = skipped_json_value.clone();
+                    action_obj["encrypt_volume"]["state"] = skipped_json_value.clone();
+                    action_obj["unmount_volume"]["state"] = skipped_json_value.clone();
+                }
+            },
+            _s => {},
+        }
+    }
+    crate::plan::write_receipt(&phase1_plan, &uninstall_args.phase1_output).await?;
+    crate::plan::write_receipt(&phase2_plan, &uninstall_args.phase2_output).await?;
+
+    Ok(())
+}
+
+fn roundtrip_to_extract_type<T: serde::de::DeserializeOwned>(
+    action: &StatefulAction<Box<dyn Action>>,
+    action_tag: &str,
+) -> eyre::Result<StatefulAction<T>> {
+    let action_json = serde_json::to_string(action).with_context(|| {
+        format!("serde_json::to_string'ing {action_tag} json to extract real type")
+    })?;
+    let action_unjson: StatefulAction<T> =
+        serde_json::from_str(&action_json).with_context(|| {
+            format!("serde_json::from_str'ing {action_tag} json to extract real type")
+        })?;
+
+    Ok(action_unjson)
```

Comment:
nit: Is there no way to get `action_tag` from `T`? It seems weird that we pass the type-to-deserialize to as a Type (`T`) and also pass `action_tag`. I was surprised to see that, but the understand it's just used for the log message. It feels redundant though if there's some compile/runtime-way to get a human readable name of `T`.

Reply:
addressed in [e5ce505](https://github.com/DeterminateSystems/nix-installer/pull/1278/commits/e5ce5053e1e1214be98c3e1c24148ad6e453f82e)